api.md

```diff
@@ -0,0 +1,215 @@
+
+### API Endpoints
+
+An "endpoint" is a combination of two things:
+
+* The verb (e.g. `GET` or `POST`)
+* The URL path (e.g. `/articles`)
+
+Information can be passed to an endpoint in either of two ways:
+
+* The URL query string (e.g. `?year=2014`)
+* HTTP headers (e.g. `X-Api-Key: my-key`)
+
+When people say "RESTful" nowadays, they really mean designing simple, intuitive endpoints that represent unique functions in the API.
+
+Generally speaking:
+
+* **Avoid single-endpoint APIs.** Don't jam multiple operations into the same endpoint with the same HTTP verb.
+* **Prioritize simplicity.** It should be easy to guess what an endpoint does by looking at the URL and HTTP verb, without needing to see a query string.
+* Endpoint URLs should advertise resources, and **avoid verbs**.
+
+Some examples of these principles in action:
+
+* [FBOpen API documentation](https://18f.github.io/fbopen/)
+* [OpenFDA example query](https://open.fda.gov/api/reference/#example-query)
+* [Sunlight Congress API methods](https://sunlightlabs.github.io/congress/#using-the-api)
+
+### Just use JSON
+
+[JSON](https://en.wikipedia.org/wiki/JSON) is an excellent, widely supported transport format, suitable for many web APIs.
+
+Supporting JSON and only JSON is a practical default for APIs, and generally reduces complexity for both the API provider and consumer.
+
+General JSON guidelines:
+
+* Responses should be **a JSON object** (not an array). Using an array to return results limits the ability to include metadata about results, and limits the API's ability to add additional top-level keys in the future.
+* **Don't use unpredictable keys**. Parsing a JSON response where keys are unpredictable (e.g. derived from data) is difficult, and adds friction for clients.
+* **Use consistent case for keys**. Whether you use `under_score` or `CamelCase` for your API keys, make sure you are consistent.
+
+### Use a consistent date format
+
+And specifically, [use ISO 8601](https://xkcd.com/1179/), in UTC.
+
+For just dates, that looks like `2013-02-27`. For full times, that's of the form `2013-02-27T10:00:00Z`.
+
+This date format is used all over the web, and puts each field in consistent order -- from least granular to most granular.
+
+
+### API Keys
+
+These standards do not take a position on whether or not to use API keys.
+
+But _if_ keys are used to manage and authenticate API access, the API should allow some sort of unauthenticated access, without keys.
+
+This allows newcomers to use and experiment with the API in demo environments and with simple `curl`/`wget`/etc. requests.
+
+Consider whether one of your product goals is to allow a certain level of normal production use of the API without enforcing advanced registration by clients.
+
+
+### Error handling
+
+Handle all errors (including otherwise uncaught exceptions) and return a data structure in the same format as the rest of the API.
```

Comment:
@higs4281, I agree that there are security implications when it comes to error handling.  For example, the sample below includes a full stack trace.  This should be avoided as it leaks implementation details to the API client.

@willbarton, I agree that appropriate HTTP status codes should be the first priority.  Bumping it up makes sense to me.

Additionally, some status codes, especially `400`, frequently get overloaded as catch-all _you did something wrong_ errors.  In these cases, it is really useful to include some programmatically accessible `code` in the response, especially if the client can take some action to resolve the error.

And for bonus points, it can also be really useful to include an `error_id`, especially for `5xx` series errors so you can correlate errors returned to clients with backend logs.